pkg/util/resourceinformer/resourceinformer.go

```diff
@@ -0,0 +1,151 @@
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	workv1alpha1 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/genericmanager"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/keys"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+	"github.com/karmada-io/karmada/pkg/util/restmapper"
+)
+
+var _ ResourceInformer = &resourceInformerImpl{}
+
+type ResourceInformer interface {
+	// BuildResourceInformers builds informer dynamically for managed resources in member cluster.
+	// The created informer watches resource change and then sync to the relevant Work object.
+	BuildResourceInformers(cluster *clusterv1alpha1.Cluster, work *workv1alpha1.Work, handler cache.ResourceEventHandler) (controllerruntime.Result, error)
+
+	// GetObjectFromCache returns object in cache by federated key.
+	GetObjectFromCache(fedKey keys.FederatedKey) (*unstructured.Unstructured, error)
+}
+
+func NewResourceInformer(c client.Client, restMapper meta.RESTMapper, informerManager genericmanager.MultiClusterInformerManager, clusterCacheSyncTimeout metav1.Duration,
+	clusterDynamicClientSetFunc func(clusterName string, client client.Client) (*util.DynamicClusterClient, error)) ResourceInformer {
+	return &resourceInformerImpl{
+		Client:                      c,
+		RESTMapper:                  restMapper,
+		InformerManager:             informerManager,
+		ClusterDynamicClientSetFunc: clusterDynamicClientSetFunc,
+		ClusterCacheSyncTimeout:     clusterCacheSyncTimeout,
+	}
+}
+
+type resourceInformerImpl struct {
+	client.Client               // used to get Cluster and Secret resources.
+	RESTMapper                  meta.RESTMapper
+	InformerManager             genericmanager.MultiClusterInformerManager
+	ClusterDynamicClientSetFunc func(clusterName string, client client.Client) (*util.DynamicClusterClient, error)
+	ClusterCacheSyncTimeout     metav1.Duration
+}
+
+// BuildResourceInformers builds informer dynamically for managed resources in member cluster.
+// The created informer watches resource change and then sync to the relevant Work object.
+func (r *resourceInformerImpl) BuildResourceInformers(cluster *clusterv1alpha1.Cluster, work *workv1alpha1.Work, handler cache.ResourceEventHandler) (controllerruntime.Result, error) {
+	err := r.registerInformersAndStart(cluster, work, handler)
+	if err != nil {
+		klog.Errorf("Failed to register informer for Work %s/%s. Error: %v.", work.GetNamespace(), work.GetName(), err)
+		return controllerruntime.Result{Requeue: true}, err
```

Comment:
Whether to "Requeue" or not, should be determined by the caller.

Reply:
Done.